### PR TITLE
feat(common): kmx file reader

### PR DESCRIPTION
**Pull request by mcdurdin**
Read a .kmx file into memory in Typescript. Currently only reads stores from the file, as these are the most common things needed for other processes. Will add support for reading other metadata and rules in future when required.

@keymanapp-test-bot skip

### CODE REVIEW

File: common/web/types/src/kmx/kmx-file-reader.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import * as r from 'restructure';
+
+export class KmxFileReader {
+  private readonly rString = new r.String(null, 'utf16le');
+
+  private readString(source: Uint8Array, offset: number): string {
+    if(offset == 0) {
+      // the file has a null string here, which is not the same as an empty
+      // string ('')
+      return null;
```
Comment (darcywong00) [inline common/web/types/src/kmx/kmx-file-reader.ts:11]: Is this ever an error to be flagged/logged?

Comment (darcywong00) [reply common/web/types/src/kmx/kmx-file-reader.ts:11]: I guess it's up to the caller to handle?

**Review by darcywong00 — APPROVED** [review]
lgtm

File: common/web/types/src/kmx/kmx-file-reader.ts (hunk 1 above)
Comment (mcdurdin) [reply common/web/types/src/kmx/kmx-file-reader.ts:11]: > I guess it's up to the caller to handle?

Yes, up to the caller. It's not an error in the file format.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.93-alpha](https://keyman.com/downloads/releases/alpha/17.0.93)

**Pull request merged**